pkg/controller/node/nodecontroller.go

```diff
@@ -474,28 +529,60 @@ func (nc *NodeController) monitorNodeStatus() error {
 
 // reconcileNodeCIDRs looks at each node and assigns it a valid CIDR
 // if it doesn't currently have one.
+// Examines the availableCIDRs set first, if no more CIDR in it, generate more.
 func (nc *NodeController) reconcileNodeCIDRs(nodes *api.NodeList) {
 	glog.V(4).Infof("Reconciling cidrs for %d nodes", len(nodes.Items))
-	// TODO(roberthbailey): This seems inefficient. Why re-calculate CIDRs
-	// on each sync period?
-	availableCIDRs := generateCIDRs(nc.clusterCIDR, len(nodes.Items))
-	for _, node := range nodes.Items {
-		if node.Spec.PodCIDR != "" {
-			glog.V(4).Infof("CIDR %s is already being used by node %s", node.Spec.PodCIDR, node.Name)
-			availableCIDRs.Delete(node.Spec.PodCIDR)
-		}
-	}
-	for _, node := range nodes.Items {
-		if node.Spec.PodCIDR == "" {
-			podCIDR, found := availableCIDRs.PopAny()
-			if !found {
-				nc.recordNodeStatusChange(&node, "CIDRNotAvailable")
-				continue
+	// check if the this node controller is restarted because of crash
+	// this will only be ran once when the controller being restarted(because of crashed) or newly start
+	if nc.needSync {
+		// if it is crashed, restore the availableCIDRs by generating CIDRs, insert them into availableCIDRs set
+		// and delete assigned CIDRs from the the availableCIDRs set
+		for _, node := range nodes.Items {
+			if node.Spec.PodCIDR != "" {
+				if nc.availableCIDRs.Has(nc.translateCIDRtoIndex(node.Spec.PodCIDR)) {
+					nc.availableCIDRs.Delete(nc.translateCIDRtoIndex(node.Spec.PodCIDR))
```

Comment:
Sorry, I am a little bit confused about the case you described. Do you mean if one Node got disconnected when we running the nc.needSync code(above code), its CIDR will not be deleted from the availableCIDRs set and may be reassign to another Node? If this is the case, it will be a little bit complicated. You are right, if we simply check the CIDR range of known nodes, it could be solved. But it will waste a lot of CIDRs for example, we have 1000 nodes at first and delete the first 999 nodes and then restart the node_controller. In this case, the first 999 CIDRs will be wasted and we will start assigning from 1001st CIDR.

The other solution will be maintaining a bool array to mark if that CIDR has been used. As you comment below, we currently has 3072 CIDRs in default setup and it will be ok to maintain a bool array(since we will use less than 1MB memory). It will both simplify the code and solve this problem, I will update the code in this way. Thanks